Mathlib/Topology/Category/LightProfinite/Limits.lean

```diff
@@ -0,0 +1,254 @@
+  desc := fun s => finiteCoproduct.desc _ fun a => s.ι.app ⟨a⟩
+  fac := fun s ⟨a⟩ => finiteCoproduct.ι_desc _ _ _
+  uniq := fun s m hm => finiteCoproduct.hom_ext _ _ _ fun a => by
+    specialize hm ⟨a⟩
+    ext t
+    apply_fun (· t) at hm
+    exact hm
+
+end FiniteCoproduct
+
+section HasPreserves
+
+instance (n : ℕ) (F : Discrete (Fin n) ⥤ LightProfinite) :
+    HasColimit (Discrete.functor (F.obj ∘ Discrete.mk) : Discrete (Fin n) ⥤ LightProfinite) where
+  exists_colimit := ⟨⟨finiteCoproduct.cofan _, finiteCoproduct.isColimit _⟩⟩
+
+instance : HasFiniteCoproducts LightProfinite where
+  out _ := { has_colimit := fun _ ↦ hasColimitOfIso Discrete.natIsoFunctor }
+
+instance {X Y B : LightProfinite} (f : X ⟶ B) (g : Y ⟶ B) : HasLimit (cospan f g) where
+  exists_limit := ⟨⟨pullback.cone f g, pullback.isLimit f g⟩⟩
+
+instance : HasPullbacks LightProfinite where
+  has_limit F := hasLimitOfIso (diagramIsoCospan F).symm
+
+noncomputable
+instance : PreservesFiniteCoproducts lightToProfinite := by
+  refine ⟨fun J hJ ↦ ⟨fun {F} ↦ ?_⟩⟩
+  suffices PreservesColimit (Discrete.functor (F.obj ∘ Discrete.mk)) lightToProfinite by
+    exact preservesColimitOfIsoDiagram _ Discrete.natIsoFunctor.symm
+  apply preservesColimitOfPreservesColimitCocone (finiteCoproduct.isColimit _)
+  exact Profinite.finiteCoproduct.isColimit _
+
+noncomputable
+instance : PreservesLimitsOfShape WalkingCospan lightToProfinite := by
+  refine ⟨fun {F} ↦ ?_⟩
+  suffices ∀ {X Y B} (f : X ⟶ B) (g : Y ⟶ B), PreservesLimit (cospan f g) lightToProfinite by
+    exact preservesLimitOfIsoDiagram _ (diagramIsoCospan F).symm
+  intro _ _ _ f g
+  apply preservesLimitOfPreservesLimitCone (pullback.isLimit f g)
+  exact (isLimitMapConePullbackConeEquiv lightToProfinite (pullback.condition f g)).symm
+    (Profinite.pullback.isLimit _ _)
+
+instance (X : LightProfinite) : Unique (X ⟶ (FintypeCat.of PUnit.{u+1}).toLightProfinite) :=
+  ⟨⟨⟨fun _ => PUnit.unit, continuous_const⟩⟩, fun _ => rfl⟩
+
+/-- A one-element space is terminal in `LightProfinite` -/
+def isTerminalPUnit : IsTerminal ((FintypeCat.of PUnit.{u+1}).toLightProfinite) :=
+  Limits.IsTerminal.ofUnique _
+
+instance : HasTerminal LightProfinite.{u} :=
+  Limits.hasTerminal_of_unique (FintypeCat.of PUnit.{u+1}).toLightProfinite
+
+/-- The isomorphism from an arbitrary terminal object of `LightProfinite` to a one-element space. -/
+noncomputable def terminalIsoPUnit :
+    ⊤_ LightProfinite.{u} ≅ (FintypeCat.of PUnit.{u+1}).toLightProfinite :=
+  terminalIsTerminal.uniqueUpToIso LightProfinite.isTerminalPUnit
+
+noncomputable instance : PreservesFiniteCoproducts LightProfinite.toTopCat.{u} where
+  preserves J _ := by
+    have : PreservesColimitsOfShape (Discrete J) Profinite.toTopCat.{u} :=
+      (inferInstance : PreservesColimitsOfShape _ (profiniteToCompHaus.{u} ⋙ compHausToTop.{u}))
+    exact (inferInstance : PreservesColimitsOfShape (Discrete J)
+      (LightProfinite.lightToProfinite.{u} ⋙ Profinite.toTopCat.{u}))
```

Comment:
In this definition and a few others, why not declaring the suitable instances for the functor `Profinite.toTopCat` as separate definitions, because it may be useful, rather than having them only inside the proofs?